### DISCUSSION TITLE
Is there a way to replace tokens within source files automatically?

### DISCUSSION THREAD

**Discussion by JayDoubleu** [Azure/azure-dev#2311] [Q&A]
I am deploying a logic App standard as part of the deployment and I have the JSON file with content shown below:
```JSON
{
    "definition": {
        "$schema": "https://schema.management.azure.com/providers/Microsoft.Logic/schemas/2016-06-01/workflowdefinition.json#",
        "actions": {
            "HTTP": {
                "inputs": {
                    "authentication": {
                        "audience": "https://cognitiveservices.azure.com",
                        "identity": "${AZURE_OPENAI_IDENTITY_RESOURCE_ID}",
                        "type": "ManagedServiceIdentity"
```

I also have my bicep file with `output AZURE_OPENAI_IDENTITY_RESOURCE_ID string = identity.outputs.resourceId`
I would like to be able to automatically replace the token within the deployment file from the output

**Comment by ellismg** [comment]
We don't have a way of doing this today.

We do have some building blocks that you could use to try to build the solution yourself.  The hooks feature would allow you run code at some point (e.g. a pre-run hook for `azd deploy`) which would let you edit this file.  `azd` should ensure that while running the hook `AZURE_OPENAI_IDENTITY_RESOURCE_ID` is defined based on the value in the `.env` file.

If we did have some built in replacement logic, I am wondering how it would work in practice.  It would be bad to edit this file in the source tree, I think, because we would then lose the replacement string.

We can get away with doing this replacement in our `.parameters.json` files because they are used as inputs to an operation and we don't have to persist the changes.  It would be harder to do that here.

**Reply by JayDoubleu** [reply]
Thanks for the reply @ellismg so while I am able to resolve these parameters within bicep itself for consumption logic apps. I have no way  to do this any more for logic apps standard nor any file consumed by azure function.

Here's how it would work in practice:

1. Deploy the platform resources via Bicep.
2. Retrieve the outputs from the Bicep deployment (`AZURE_OPENAI_IDENTITY_RESOURCE_ID` is one of the outputs).
3. Replace all tokens in the `src/` directory (in this case, replacing the AZURE_OPENAI_IDENTITY_RESOURCE_ID value in one of the JSON files).
4. Deploy the Logic App onto the App Service (after the files have been zipped with all tokens replaced).

This is a significant gap that we, as the platform engineering teams, would typically fill using Azure DevOps pipelines.

To put it another (simplified) way :

1. We would have "platform" Bicep resources in one directory.
2. We would keep the function/Logic App/Web App code in a second directory.
3. Azure DevOps pipelines would pull from the repository and perform the same logic as explained in steps 1 through 4.
4. Step 3 above would usually be accomplished using a number of "wrappers" in combination with YAML Azure DevOps pipelines.

Unfortunately, there is no way to replicate this process locally on your laptop.

This is where I see `azd` stepping in and saving the day, allowing us to avoid using any PowerShell or general shell wrappers. It generally fills the gap between Platform deployments and zip deployments to the App Service.
